src/Logary/MessageModule.fs

```diff
@@ -186,14 +186,41 @@ module Message =
   let hasTag (tag: string) (message: Message) =
     message |> getAllTags |> Set.contains tag
 
+  /// A single Message can take multiple gauges; use this function to add further
+  /// gauges to the message. You can add gauges to events as well.
+  [<CompiledName "AddGauge">]
+  let addGauge gaugeName (gauge: Gauge) message =
+    let gaugeName = KnownLiterals.GaugeNamePrefix + gaugeName
+    message |> setContext gaugeName gauge |> tag KnownLiterals.GaugeTag
+
+  /// A single Message can take multiple gauges; use this function to add further
+  /// gauges to the message. You can add gauges to events as well.
+  [<CompiledName "AddGauges">]
+  let addGauges (gauges: #seq<string * Gauge>) message =
+    (message, gauges) ||> Seq.fold (fun m (n, g) -> m |> addGauge n g)
+
+  /// if one log message needs metric its count, then we can invoke this method for shortcut.
+  /// if one needs metric gauges or histogram, it's better to define a metric type to handle this case.
+  /// since these metric are not bound to message/logger, and can be cached by filed declaration (more efficient).
+  /// labelNames will grab its value from message's context: fields -> contexts, if not found, will raise error
+  [<CompiledName "EnableCounterMetric">]
+  let enableCounterMetric (counterConf: Metric.BasicConf) (message: Message) =
```

Comment:
hmmmmmm....

I don't like this design TBH... Do we really have to store it in the Message...?

Reply:
```f#

module A =
  open Logary.Metric

  let logger = Log.create "xxx.yyy"

  let parseConfigFailedCounter =
    GaugeConf.create {name="xxx_yyy_parse_config_faild_total"; description= "the total number for xxx.yyy.parseConfig faild"; labelNames = [| "path" |]}
    |> LogManager.DefaultMetricRegistry.registerMetric

  // add gauge metric and histogram with buckets [| 0.001; 0.005; 0.01; 0.025; 0.05; 0.1; 1.0|] in the same time
  let parseConfigGaugeWithHistogram =
    GaugeConf.create {name="xxx_yyy_parse_config_latency_seconds"; description= "xxx.yyy.parseConfig latency in seconds"; labelNames = [||]}
    |> GaugeConf.withHistogram [| 0.001; 0.005; 0.01; 0.025; 0.05; 0.1; 1.0|]
    |> LogManager.DefaultMetricRegistry.registerMetric
    |> Metric.noLabels


  // using default buckets:  [| 0.005; 0.01; 0.025; 0.05; 0.075; 0.1; 0.25; 0.5; 0.75; 1.; 2.5; 5.0; 7.5; 10. |]
  let parseConfigLatencyHistogram =
    HistogramConf.create("xxx_yyy_parse_config_latency_seconds", "xxx.yyy.parseConfig latency in seconds")
    |> LogManager.DefaultMetricRegistry.registerMetric
    |> Metric.noLabels

  // using hookup and shortcut style
  let parseConfigWithShortcut path =
    try
      // if we hook up metric at logary's conf, then there is no need to do other things here.
      // span will be treated as gauges when span has been finished.
      use traceSpan =
        logger.buildSpan()
        |> Span.setMessage (eventX "time usage about parse config with {path}" >> setField "path" path)
        |> Span.start

      // do some work here
      failwith "not implement"
      
    with ex ->
      eventX "parse config from {path} failed"
      >> setField "path" path
      >> enableCounterMetric {name="xxx_yyy_parse_config_faild_total"; description= "the total number for xxx.yyy.parseConfig faild"; labelNames = [| "path" |]}
      |> logger.error

  // using metric as a separate instance style
  let parseConfig path =
    try
      let traceSpan =
        logger.buildSpan()
        |> Span.setMessage (eventX "time usage about parse config with {path}" >> setField "path" path)
        |> Span.start

      // do some work here
      failwith "not implement"


      let spanLog = traceSpan.finish id
      // gauge with histogram like below
      parseConfigGaugeWithHistogram.set (Duration.FromTicks(spanLog.duration).TotalSeconds)
      // or metric only histogram like below
      parseConfigLatencyHistogram.observe (Duration.FromTicks(spanLog.duration).TotalSeconds)

    with ex ->
      eventX "parse config from {path} failed"
      >> setField "path" path
      |> logger.error

      (parseConfigFailedCounter.labels [| path |]).inc 1.

```

`enableCounterMetric` is just some shortcut for creating a counter metric for a log message (means its occurrence total count, if this message is not send because its rules or logging level's config, this counter will not metric either), we can use `metric` separately like code above `parseConfig` . We recommend that use metric separately, even if the message are not send, the counter still metric. and using metric separately will not have to do a find operation to get the real metric, can save some time. 

> These should be primarily used as file-static variables, that is, global variables defined in the same file as the code they’re instrumenting. The client library SHOULD enable this. The common use case is instrumenting a piece of code overall, not a piece of code in the context of one instance of an object. Users shouldn’t have to worry about plumbing their metrics throughout their code, the client library should do that for them (and if it doesn’t, users will write a wrapper around the library to make it "easier" - which rarely tends to go well).

says from https://prometheus.io/docs/instrumenting/writing_clientlibs/#metrics

Reply:
or do you have any ideas how we can design a api to wrapper these inside ?
prometheus's metric needs:
- metric name 
- metric help (optional)
- labels (`Map<string,string>`) (optional, depend on use case)

each metric with same labels needs use the same metric, 
```f#
let metric = Metric("metricA").withLabelName("http_verb")
metric.label("get").inc 5
metric.label("get").inc ()
metric.label("post").inc ()
```
code above need export metric as:
metricA{http_verb=get} 6
metricA{http_verb=post} 1

Reply:
I see your points, but I don't think that what you've described is an optimal solution. If you commit that file in here, I can see if I can make the API more to my liking in this PR

Reply:
`gauge "metric_a" 6 |> setField "http_verb" "get"`

Reply:
In your exporter then; `exporter |> varyBy "http_verb"`

Reply:
or even default the exporter to vary always for every Map of field it sees.

Reply:
In fact, I don't recommend such a way (`enableCounterMetric`) to bind messages(prefer remove this function), one should add a metric type separately instead of using this shortcut. Then no ship configuration in the log message.

i think this may some misunderstand here, "global style" it means use the global `DefaultMetricRegistry` in `LogManager`, but if user can grab a `LogManager` instance, then they can use logManager's `metricRegistry` to register these metrics (not the global one, the one in logManager's instance).

and the `enableCounterMetric` style is used with `logManager's instance style`, since metric conf is bind with a log message in this way, a log message will be send through a logManager(logary `Registry`), at that time it hook up the `metricRegistry` at that logary's registry instance. **This just save time for define another separate metric**. but this will `register/find` the metric on the fly (when sending log message), **which is not recommend** through https://prometheus.io/docs/instrumenting/writing_clientlibs/#metrics.